packages/react-query/src/__tests__/prefetch.test.tsx

```diff
@@ -0,0 +1,353 @@
+import { describe, expect, it, vi } from 'vitest'
+import React from 'react'
+import { fireEvent, waitFor } from '@testing-library/react'
+import {
+  QueryCache,
+  usePrefetchInfiniteQuery,
+  usePrefetchQuery,
+  useSuspenseInfiniteQuery,
+  useSuspenseQuery,
+} from '..'
+import { createQueryClient, queryKey, renderWithClient, sleep } from './utils'
+
+describe('usePrefetchQuery', () => {
+  const queryCache = new QueryCache()
+  const queryClient = createQueryClient({ queryCache })
+
+  it('should prefetch query if query state does not exist', async () => {
+    const queryOpts = {
+      queryKey: queryKey(),
+      queryFn: vi.fn().mockImplementation(async () => {
+        await sleep(10)
+
+        return 'Prefetch is nice!'
+      }),
+    }
+
+    const anotherSpy = vi.fn()
+
+    function Suspended() {
+      const state = useSuspenseQuery({ ...queryOpts, queryFn: anotherSpy })
+
+      return (
+        <div>
+          <div>data: {String(state.data)}</div>
+          <div>fetching: {state.isFetching ? 'true' : 'false'}</div>
+        </div>
+      )
+    }
+
+    function App() {
+      usePrefetchQuery(queryOpts)
+
+      return (
+        <React.Suspense fallback="Loading...">
+          <Suspended />
+        </React.Suspense>
+      )
+    }
+
+    const rendered = renderWithClient(queryClient, <App />)
+
+    await waitFor(() => rendered.getByText('data: Prefetch is nice!'))
+    expect(rendered.queryByText('fetching: true')).not.toBeInTheDocument()
+    expect(queryOpts.queryFn).toHaveBeenCalledTimes(1)
+    expect(anotherSpy).not.toHaveBeenCalled()
+  })
+
+  it('should not prefetch query if query state exists', async () => {
+    const queryOpts = {
+      queryKey: queryKey(),
+      queryFn: vi.fn().mockImplementation(async () => {
+        await sleep(10)
+
+        return 'Prefetch is really nice!'
+      }),
+    }
+
+    const anotherSpy = vi.fn()
+
+    function Suspended() {
+      const state = useSuspenseQuery({ ...queryOpts, queryFn: anotherSpy })
+
+      return (
+        <div>
+          <div>data: {String(state.data)}</div>
+          <div>fetching: {state.isFetching ? 'true' : 'false'}</div>
+        </div>
+      )
+    }
+
+    function App() {
+      usePrefetchQuery({ ...queryOpts, queryFn: anotherSpy })
+
+      return (
+        <React.Suspense fallback="Loading...">
+          <Suspended />
+        </React.Suspense>
+      )
+    }
+
+    await queryClient.fetchQuery(queryOpts)
```

Comment:
can we please add another test where we have a failed query in the cache? This used to bite us when the implementation was still `ensureQueryData` - the query was there, but data was not.

I think you can get that with:


```
await queryClient.prefetchQuery({
  ...queryOpts,
  queryFn: () => { 
    throw new Error("oops")
  },
  retry: false
})
```

prefetching doesn't throw so there shouldn't be any try/catch needed.

Reply:
Of course, doing it right now

Reply:
Added two error scenarios in the test file